--- ../src-base/minecraft/cpw/mods/fml/common/registry/LanguageRegistry.java
+++ ../src-work/minecraft/cpw/mods/fml/common/registry/LanguageRegistry.java
@@ -287,10 +287,12 @@
                 added.add(lang);
                 LanguageRegistry.instance().injectLanguage(lang, StringTranslate.parseLangFile(zf.getInputStream(ze)));
                 // Ensure en_US is available to StringTranslate on the server
+                /*
                 if ("en_US".equals(lang) && side == Side.SERVER)
                 {
                     StringTranslate.inject(zf.getInputStream(ze));
                 }
+                */
             }
         }
         if (added.size() > 0)
@@ -314,6 +316,7 @@
                 FMLLog.fine("Injecting found translation assets for lang %s at %s into language system", lang, currPath);
                 LanguageRegistry.instance().injectLanguage(lang, StringTranslate.parseLangFile(new FileInputStream(file)));
                 // Ensure en_US is available to StringTranslate on the server
+
                 if ("en_US".equals(lang) && side == Side.SERVER)
                 {
                     StringTranslate.inject(new FileInputStream(file));
