--- ../src-base/minecraft/net/minecraft/network/EnumConnectionState.java
+++ ../src-work/minecraft/net/minecraft/network/EnumConnectionState.java
@@ -1,5 +1,6 @@
 package net.minecraft.network;
 
+import cc.uraniummc.packet.S45PacketTitle;
 import com.google.common.collect.BiMap;
 import com.google.common.collect.HashBiMap;
 import com.google.common.collect.Iterables;
@@ -181,6 +182,7 @@
             this.func_150756_b(62, S3EPacketTeams.class);
             this.func_150756_b(63, S3FPacketCustomPayload.class);
             this.func_150756_b(64, S40PacketDisconnect.class);
+            this.func_150756_b(69, S45PacketTitle.class);
             this.func_150751_a(0, C00PacketKeepAlive.class);
             this.func_150751_a(1, C01PacketChatMessage.class);
             this.func_150751_a(2, C02PacketUseEntity.class);
