--- ../src-base/minecraft/net/minecraft/network/play/server/S02PacketChat.java
+++ ../src-work/minecraft/net/minecraft/network/play/server/S02PacketChat.java
@@ -3,6 +3,8 @@
 import cpw.mods.fml.relauncher.Side;
 import cpw.mods.fml.relauncher.SideOnly;
 import java.io.IOException;
+
+import lombok.Getter;
 import net.minecraft.network.INetHandler;
 import net.minecraft.network.Packet;
 import net.minecraft.network.PacketBuffer;
@@ -14,7 +16,6 @@
     private IChatComponent field_148919_a;
     private boolean field_148918_b;
     private static final String __OBFID = "CL_00001289";
-
     public S02PacketChat()
     {
         this.field_148918_b = true;
@@ -27,7 +28,6 @@
 
     public S02PacketChat(IChatComponent p_i45180_1_, boolean p_i45180_2_)
     {
-        this.field_148918_b = true;
         this.field_148919_a = p_i45180_1_;
         this.field_148918_b = p_i45180_2_;
     }
